TypeCobol.LanguageServer/TypeCobolServer.cs

```diff
@@ -89,27 +89,73 @@ private IEnumerable<CodeElementWrapper> CodeElementFinder(FileCompiler fileCompi
             int linesCount = fileCompiler.CompilationResultsForProgram.ProgramClassDocumentSnapshot.PreviousStepSnapshot.Lines.Count;
             if (linesCount != 0 && lineIndex < linesCount)
             {
+                bool isFirstLine = true;
                 while (codeElements.Count == 0 && lineIndex >= 0)
                 {
                     var codeElementsLine =
-                        fileCompiler.CompilationResultsForProgram.ProgramClassDocumentSnapshot.PreviousStepSnapshot.Lines[lineIndex];
+                        fileCompiler.CompilationResultsForProgram.ProgramClassDocumentSnapshot.PreviousStepSnapshot
```

Comment:
Can you explain in why the problem happens for WHEN ?
Can the problem happens for other statements?

Reply:
You write a perform with when statements.
When you write only a new WHEN and Ctrl+Space, the completion searchs the nearest codeElement (it can be display, perform, ...).
The problem is that the WHEN is not identified as a codeElement but only as token.
Solution is to detect WHEN on the line of Ctrl+Space if there is no codeElement on this line.

Reply:
Ok, in this case, we should rather look at the grammar of WHEN to allow a WHEN without anything after it.

Reply:
So after your investigation and our discussions:
- We must change the grammar to accept a WHEN without anything else after. This ensure the WhenCondition codeElement is created
- It's also pretty sure we need to change the grammar of Evaluate/When/End evaluete (CUPS grammar) to accept a WHEN not followed by any statements.

By doing this, we'll have proper CodeElement and Node.

But this means our grammar is too permissive and we need to check the syntax manually.
@mayanje @fm-117 @LANOYDO @rooksdo  What do you think ?

I think it's better to change the grammar because we can then rely on CodeElement and Node.
Otherwise we would have to deal with tokens only in LanguageServer. That means try to do the job of the parser to deal with the token WHEN.

Reply:
This is somewhat close to #1440 where parsing is interrupted and no completion data can be found. I still think that the grammar should be as strict as possible but the recovery mechanism should be improved to allow such missing parts. Improving the parser recovery (ANTLR and CUPS) would be useful for all completion scenarios not only the WHEN case.
Easir said than done of course...

Reply:
I think we first need to investigate why for an incomplete move instruction we still have a move CodeElement and Node.

In following code, `move` will be detected and we even have an error because `var1` is not referenced.
```cobol
       IDENTIFICATION DIVISION.    
       PROGRAM-ID. DVZZBCO1.
       PROCEDURE DIVISION.
           move var1
           .

       END PROGRAM DVZZBCO1.
```

Reply:
move var1
.
sentence-0
        [[MoveStatement]] [12,15:move]<MOVE> --> [17,20:var1]<UserDefinedWord>
        end

move 
.
 sentence-0
        [[MoveStatement]] No Consumed Tokens
        [[MoveStatement]] [12,15:move]<MOVE> --> [12,15:move]<MOVE>
        end

-------- with non modified code
evaluate true
     when bool1
end-evaluate
.
 [[EvaluateStatement]] [12,19:evaluate]<EVALUATE> --> [21,24:true]<TRUE>


evaluate true
     when 
end-evaluate
.
sentence-0
        [[EvaluateStatement]] [12,19:evaluate]<EVALUATE> --> [21,24:true]<TRUE>

          end
        end

-------- with modified code (issue1593/PR1610)
evaluate true
     when bool1
end-evaluate
.
      [[EvaluateStatement]] [12,19:evaluate]<EVALUATE> --> [21,24:true]<TRUE>

        ?
          [[WhenCondition]] [17,20:when]<WHEN> --> [22,26:bool1]<UserDefinedWord>

        ?
          end
        end

evaluate true
     when 
end-evaluate
.
      [[EvaluateStatement]] [12,19:evaluate]<EVALUATE> --> [21,24:true]<TRUE>
        ?
          [[WhenDummy]] [17,20:when]<WHEN> --> [17,20:when]<WHEN>

        ?
          end
        end